File: python/Neetcode/TwoPointers/Easy/680.ValidPalindrome.py
# Given a string s, return true if the s can be palindrome after 
# deleting at most one character from it.

def validPalindrome(s):
    left, right = 0, len(s) - 1

    while left < right:
        if s[left] != s[right]:
            l_remove, r_remove = s[left + 1:right + 1], s[left:right]
            return l_remove == l_remove[::-1] or r_remove == r_remove[::-1]
        
        left, right = left + 1, right - 1

    return True

print(validPalindrome("aba"))
print(validPalindrome("abca"))
print(validPalindrome("abc"))

File: swift/Neetcode/TwoPointers/Easy/680.ValidPalindrome.swift
// Given a string s, return true if the s can be palindrome after 
// deleting at most one character from it.

func validPalindrome(_ s: String) -> Bool {
    let s = Array(s)
    var left = 0
    var right = s.count - 1

    while left < right {
        if s[left] != s[right] {
            let lRemove = String(s[(left + 1)...right])
            let rRemove = String(s[left..<(right)])
            
            return lRemove == String(lRemove.reversed()) || rRemove == String(rRemove.reversed())
        }
        
        left += 1
        right -= 1
    }
    
    return true
}

print(validPalindrome("aba"))
print(validPalindrome("abca"))
print(validPalindrome("abc"))